File: leetcode-new/131.cpp
class Solution
{
public:
    vector<vector<string>> partition(string s)
    {
        vector<vector<string>> res;
        vector<string> cand;
        dfs(res, s, cand, 0);
        return res;
    }

    bool isPalindrome(string str)
    {
        int l = 0;
        int r = str.length() - 1;
        while (l < r)
        {
            if (str[l] != str[r])
            {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }

    void dfs(vector<vector<string>> &res, const string &s, vector<string> &cand, int i)
    {
        if (i == s.length())
        {
            res.push_back(cand);
            return;
        }
        for (int j = i + 1; j <= s.length(); j++)
        {
            string neww = s.substr(i, j - i);
            if (isPalindrome(neww))
            {
                cand.push_back(neww);
                dfs(res, s, cand, j);
                cand.pop_back();
            }
        }
    }
};

File: leetcode-venture/131.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        def dfs(i, cand):
            if i == len(s):
                res.append(cand)
                return

            for j in range(i, len(s)):
                if self.isPalindrome(s, i, j):
                    dfs(j+1, cand+[s[i:j+1]])
            
            return

        dfs(0, [])
        return res

    def isPalindrome(self, s, i, j):
        while i < j:
            if s[i] == s[j]:
                i += 1
                j -= 1
            else:
                return False
            
        return True
